### ISSUE TITLE
Planning for the inclusion of image recognition metadata

### ISSUE THREAD

**Issue by kitenetter** [BiologicalRecordsCentre/iRecord#1786]
Notes from John at 16 Jan 2025:

Adding key information about classification results into the data kept for a record in Elasticsearch and updating the filter tool (if required) will allow verifiers to quickly overview classification information for a set of records, filter to those which have a high or low degree of confidence and act accordingly (e.g. bulk verify [or bulk redet]). Suggested fields include:
- Probability assigned by the classifier to the current determination for the record, only if the current determination matches one of the classifier suggestions given. This is a number between 0 and 1 (or could be expressed as a percentage up to 100). It would be 0 if the record was classified but a different determination chosen, or empty if there was no classification event (or at least, no metadata stored).
- Is the current determination the one chosen by the human from the list given by the classifier? True or false (empty if not classified).
- Is the current determination the one chosen by the classifier as the most likely? True or false (empty if not classified).

[John: I have added the word "likely" in the last line above, is that what you meant?]

When a verifier wants to drill into the details of a single record, a new “Image classifier” tab on the details pane can show full information. This would include a list of classification events (normally only 1) with the date, time, classifier used, version, then a list of suggestions and probabilities. Icons would be used to indicate which suggestion was chosen as most likely by the classifier, plus which was chosen by the recorder. Optionally this tab could show the actual images used, as it is possible that a new image was added to the record or removed from the record after classification so those used may not exactly match the current list.

Additional notes from me:
Classifier scores: for the iRecord app we are moving toward a three-category 'traffic-light' approach, rather than giving the actual percentage number, mainly because the % numbers don't equate to a level of confidence in a straightforward way. So we might want to take the same approach here (although it might also be possible to give both the traffic-light summary and the detailed %).

Additional tab: I think will probably be necessary to display the full range of detail as you suggest, but I would also like to have the most important elements displayed within the 'photo cards' for each occurrence (on the photo version of the verification page), along the lines of the mock-up we previously looked at:

![Image](https://github.com/user-attachments/assets/4ed8df31-2a7b-4fe0-8243-58c669483096)

Exactly what the most important elements are remains to be defined.

Note that as yet we don't collect classifier metadata from the iRecord app but this is [being progressed](https://github.com/NERC-CEH/irecord-app/issues/297).

**Comment by sacrevert** [comment]
Does the second bullet imply that the human looked at the list or not? For example, in the app, if I choose a species name before taking photos, then take photos that get classified, if the classifier lists my name, does that result in TRUE in this field even though I didn't select it from the list? It's not clear to me whether this field is merely about whether or not the current det matches something in the list, or whether it implies that some specific human action/click has been captured (i.e the fact that the human originally looked at the classifier list or not). If it's not about agency, is it possible to capture whether the human selection was actually from a classifier-enhanced list or not?

Also, my preference would be not to expose the uncalibrated model probabilities to verifiers.

**Comment by johnvanbreda** [comment]
@sacrevert I'm not aware that we can capture if the user selected from the classifier list vs choosing a name independently of the classifier. I guess this is a question for the app. My intention was to simply flag situations where the user and classifier completely disagree, which I think is useful information regarding record quality to highlight to the verifier (e.g. they may choose to spend a session looking at problem records where there is complete disagreement, so could easily filter on this field). Your suggestion would allow us to work out how often the classifier is actually being used to pick a species for the record  which is perhaps more about justifying whether the classifier is actually useful? Or did you think this would be useful information during verification?

**Comment by johnvanbreda** [comment]
@kazlauskis a quick question - does the iRecord app (or any other app using classifiers) populate the `classification_suggestions.human_chosen` field with true _only when the user chooses a species from the list of classifier responses_ or does this field get set if the species selected matches the classifier one even if the user manually entered the species name and didn't pick from the classifier results?

**Comment by kitenetter** [comment]
I believe we do capture the information on how a name was chosen - Jim told me:

> In the Indicia database occurrences table there is a field called machine_involvement which "Identifies the involvement of machine image classifiers in the identification”:
> •	  Null = unknown;
> •	  0 = no involvement;
> •	  1 = human determined, machine suggestions were ignored;
> •	  2 = human chose a machine suggestion given a very low probability;
> •	  3 = human chose a machine suggestion that was less-preferred;
> •	  4 = human chose a machine suggestion that was the preferred choice;
> •	  5 = machine determined with no human involvement."
> 
> I've queried the database and find there are 26k records where this field has been set. They are records for EBMS (19k) and E-Surveyor (7.5k) but not iRecord .

Those notes are extracted from a longer email that Jim sent, see:

[Image classifier data captured - Jim Bacon.docx](https://github.com/user-attachments/files/18708085/Image.classifier.data.captured.-.Jim.Bacon.docx)

**Comment by sacrevert** [comment]
@johnvanbreda said 
>Your suggestion would allow us to work out how often the classifier is actually being used to pick a species for the record which is perhaps more about justifying whether the classifier is actually useful? Or did you think this would be useful information during verification?

Primarily I was thinking about the first, especially where folk are using these data for research questions, but your second point is reasonable: I imagine it would be useful there too sometimes.

**Comment by kazlauskis** [comment]
> @kazlauskis a quick question - does the iRecord app (or any other app using classifiers) populate the classification_suggestions.human_chosen field with true only when the user chooses a species from the list of classifier responses or does this field get set if the species selected matches the classifier one even if the user manually entered the species name and didn't pick from the classifier results?

@johnvanbreda, I think we set it even if the user has selected the species manually without AI's help.

**Comment by johnvanbreda** [comment]
Coding now done in the develop branch.

**Comment by johnvanbreda** [comment]
This is now on the main verification and photo verification page on the test version of iRecord for you to check @kitenetter. Note that on iRecord there are only a few training records which actually have any image classification metadata at the moment.

In order to enable this feature on live, the following setting needs to be added to the verification tool's recordDetails and cardGallery control:
```
@includeImageClassifierInfo=true
```

**Comment by kitenetter** [comment]
@johnvanbreda so far I've been unable to find any training occurrences that have classifier metadata, can you give me an example or two, or let me know if there is a way of filtering for them?

**Comment by johnvanbreda** [comment]
I may have been mistaken, or the records have been since deleted. At the moment there are only records in the e-Surveyor and EBMS datasets that have classifier data so they are not available in iRecord.

**Comment by kitenetter** [comment]
@johnvanbreda I think we do now have some iRecord app records that have the classifier data. Is the test page still available, and if so can you remind me of the URL?

**Comment by kitenetter** [comment]
@johnvanbreda bumping this one - is the test page available?

**Comment by johnvanbreda** [comment]
@kitenetter I've added it to the test version of iRecord, on the main photo verification page for you to try. Note that this is the first time I've seen it with real data, so I've just made a couple of amendments that will only appear next time we do an update:
1. Round the percentage probability shown for suggestions to an integer.
2. Hide the classifier name and version for the version shown on the card gallery (it will still show on the classification info tab).

**Comment by sacrevert** [comment]
This may not be much of an issue for macromoths where all species are well-represented in the classifier training data and likely also represent the range of qualities of photos that users will submit, but for other groups, we should present some sort of warning to the user that the probabilities are _not_ in general $$\mathrm{Pr}(speciesX)$$ but actually $$\mathrm{Pr}(speciesX | trainingData)$$. If the training set is incomplete or poorly represents submitted photos (e.g. classifier trained on perfect photos, but submitted photos normally low quality), then these probabilities are not the same. This is why we switched to discretised traffic-light scale feedback on the app.

**Comment by johnvanbreda** [comment]
@sacrevert understood - I'll let @kitenetter review the current display before making any changes though.

**Comment by kitenetter** [comment]
@johnvanbreda I'm looking at this page:
https://test-brc-irecord-d10.pantheonsite.io/verification/photos

and I've filtered for macro-moths added via the iRecord app. So far I can't find any display of classifier info - am I looking in the wrong place?

**Comment by johnvanbreda** [comment]
I suspect the site database had been copied from live at some point as the configuration option to enable this was missing. I've added it. The Quality panel on the filter builder allows you to filter for any record with a classification added so it's easy to find them.

**Comment by kitenetter** [comment]
This looks great.

Initial comments:
1. Do we actually need the "Image classifier info" tab? Currently it isn't adding anything to the information provided on the 'card' for the photo record.
2. Alternatively, could we lose the "Nature Identification API" metadata from the photo card display, and retain it in the "Image classifier info" tab? That would reduce the space taken up on the photo card, and I doubt that the API metadata will need to be checked every time.
3. When you click on a classifier suggestion in order to redetermine the record, the redet popup opens - it would be good if the "Redetermine to" box could be auto-filled with the taxon you have clicked on,
4. We don't need the full set of decimal places for the classifier confidence score, and in fact I think we could just use the three traffic light options rather than showing the percentage figure.
5. Suggest reducing the length of "Image classifier conflicts with identification provided." Would this be clear enough: "Classifier conflicts with ID." Or maybe just use an icon to indicate disagreement, with the text as a hover-over.
6. Currently we are picking up classifier metadata from individual app records (survey ID 374, "iRecord App"), but not from survey-list records (survey ID 576, "iRecord App General Survey"). Is that configurable by you or does Karolis need to change anything at the app end?
7. I expect you intend this anyway, but in due course it would be good to enable the "Image classifier info" tab on the main verification page as well as the photo verification page.

**Comment by johnvanbreda** [comment]
1 & 2 - newer versions of this code will indeed hide the classifier name and version from the info shown on the main photo card, so you have to look at the details in the Image classifier info tab to see this. We need a tab as that's the only way to see the info on the main no-photo verification page.
3. This will be fixed in the next code release.
4. Agreed and done for the next release. There will be a hover-tooltip with the rounded percentage number.
5. Agreed and done for the next release. It now shows an icon with a hover tooltip containing the text that's currently displayed as a label.
6. I think that's app-end, I don't do anything survey specific.
7. Agreed, will do that when deploying the final version.

Here's an example of the new look - there are hover hints over the tick in the heading, plus to tell you the probabilities on each suggestion's circle badge:

<img width="403" height="163" alt="Image" src="https://github.com/user-attachments/assets/8193d020-0037-4e0f-8ec0-6150fd8e85cc" />

**Comment by kitenetter** [comment]
Re point 6, I can in fact now see classifier metadata appearing for survey ID 576, "iRecord App General Survey", so either something has changed or else I was looking in the wrong place before. In any case it now seems to be working as expected.

**Comment by johnvanbreda** [comment]
@kitenetter shall I enable this feature on the live site yet?

**Comment by kitenetter** [comment]
@johnvanbreda I can no longer see the classifier metadata on the test site (https://test-brc-irecord-d10.pantheonsite.io/verification/photos), has something been over-written?

**Comment by kitenetter** [comment]
@johnvanbreda I'll need to be able to test and take screenshots this week for NCEA reporting, can we enable this either on the test site or on the live site?

**Comment by johnvanbreda** [comment]
I've added the feature back onto https://test-brc-irecord-d10.pantheonsite.io/verification/photos @kitenetter. If you can confirm that the functionality is OK I can put it onto the main live verification pages so it's not lost again.

**Comment by kitenetter** [comment]
@johnvanbreda it looks like I've again left it too long to be able to test this - the verification page on the test site has some layout problems, and I can't find any occurrences that have classifier metadata showing. Does the test site need to be updated again?
